src/ast/document-descriptor.ts

```diff
@@ -38,3 +46,206 @@ export class DocumentDescriptor implements Descriptor {
     return this.document.url;
   }
 }
+
+export interface Feature {
+  kinds: Iterable<string>;
+  identifiers?: Iterable<string>;
+  // sourceLocation: SourceLocation;
+}
+
+
+export interface Resolvable extends ScannedFeature {
+  resolve(document: Document): Feature;
+}
+function isResolvable(x: any): x is Resolvable {
+  return x.resolve && typeof x.resolve === 'function';
+}
+
+export class Document implements Feature {
+  url: string;
+  parsedDocument: ParsedDocument<any, any>;
+  isInline: boolean;
+
+  kinds: Iterable<string>;
+  identifiers: Set<string>;
+
+  private _rootDocument: Document;
+  // Should be able to emulate this once every Feature has a SourceLocation by
+  // just filtering out by feature.sourceLocation.file === this.url;
+  private _localFeatures = new Set<Feature>();
+
+  static makeRootDocument(scannedDocument: ScannedDocument): Document {
+    const result = new Document(scannedDocument);
```

Comment:
This needs some comments. It's not immediately clear why a document is added as a feature of itself.